Add docs for JS (#39)
SAP/ai-sdk: docs-js/orchestration/chat-completion.mdx

---
id: chat-completion
title: Chat Completion
hide_title: false
hide_table_of_contents: false
description: How to use the SAP Cloud SDK for AI to perform chat completion tasks using orchestration service from SAP AI Core.
keywords:
  - sap
  - cloud
  - sdk
  - ai
  - orchestration
  - chat
  - completion
---

The `@sap-ai-sdk/orchestration` package provides a client for the orchestration service of SAP AI Core.
The orchestration service harmonizes the API of various generative AI models, enabling seamless integration and interaction with different models through a unified interface.
Additionally, it provides features like [templating](#templating), [content filtering](#content-filtering), [grounding](#grounding) and many more to enhance the interaction with generative AI models.

Find more details about orchestration workflow [here](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/orchestration-workflow).

## Installation

```
$ npm install @sap-ai-sdk/orchestration
```

## Usage

> Initialize a client with proper orchestration configuration for each modules.

Reviewer: ```suggestion
Initialize a client with proper orchestration configuration for each module.
```

Below is an example for configuring the mandatory `llm` and `templating` modules.

```ts
import { OrchestrationClient } from '@sap-ai-sdk/orchestration';

const orchestrationClient = new OrchestrationClient({
  llm: {
    model_name: 'gpt-4o'
  },
  templating: {
    template: [
      { role: 'user', content: 'Hello World! Why is this phrase so famous?' }
    ]
  }
});
```

> Here we specify the model name with a user message as prompt.

Reviewer: ```suggestion
Here, the model name is specified along with a user message as the prompt.
```

Its not a recommended practice to write instructional docs using first person. if really needed, use `you` (even though its often implied with imperative case)

To send a chat completion request, use the `chatCompletion()` method.
Use the following convenience methods for handling chat completion response:

- `getContent()` parses the response and returns the model's output as a string.
- `getFinishReason()` retrieves the `finish_reason` explaining why chat completion request stopped.
- `getTokenUsage()` provides token usage details, including `total_tokens`, `prompt_tokens`, and `completion_tokens`.

```ts
const response = await orchestrationClient.chatCompletion();

console.log(response.getContent());
console.log(response.getFinishReason());
console.log(JSON.stringify(response.getTokenUsage()));
```

> In the following sections, we explain the configuration for each module in detail.

Reviewer: ```suggestion
The following sections provide detailed configuration for each module.
```

> In the following sections, we explain the configuration for each module in detail.

Reviewer: I feel like we dont need these lines here anymore. With docusaurus we have links displayed on the right which clearly shows docs structure. I would also add the sample code link somewhere else (maybe at the top)

Author: Currently heading starting from the third level are not displayed on the right sidebar. Not sure if we can configure this.

Author: I moved the sample code link before the first example.

In addition, you can find more **sample code** [here](https://github.com/SAP/ai-sdk-js/blob/main/sample-code/src/orchestration.ts).

### LLM Configuration

Choose the LLM by setting the `model_name` property in `llm` module configuration.
Optionally, define `model_version` (default: `latest`) and `model_params` for custom settings.

```ts
const llm: LlmModuleConfig = {
  model_name: 'gpt-4o',
  model_version: '2024-08-06', // optional
  model_params: {
    // optional
    max_tokens: 50,
    temperature: 0.1
  }
};
```

#### Harmonized API

The orchestration service provides a [harmonized API](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/harmonized-api) for all models.

> It lets you use different models with the same API.

Reviewer: ```suggestion
It allows using different models with the same API.
```

The example below returns a JSON Schema with `strict` set to `true` to let the response adhere to the schema definition.

```ts
const templating: TemplateModuleConfig = {
  template,
  response_format: {
    type: 'json_schema',
    json_schema: {
      name: 'capital_response',
      strict: true,
      schema: {
        type: 'object',
        properties: {
          country_name: {
            type: 'string',
            description: 'The name of the country provided by the user.'
          },
          capital: {
            type: 'string',
            description: 'The capital city of the country.'
          }
        },
        required: ['country_name', 'capital']
      }
    }
  }
};
```

You can also define JSON schema using [Zod](https://zod.dev/) schema as shown below:

```ts
const countryCapitalSchema = z
  .object({
    country_name: z.string(),
    capital: z.string()
  })
  .strict();

const response_format: ResponseFormatJsonSchema = {
  type: 'json_schema',
  json_schema: {
    name: 'capital_response',
    strict: true,
    schema: zodToJsonSchema(countryCapitalSchema)
  }
};
```

### Prompt Registry

> While you can define template content as part of the templating configuration, you can also refer a template defined in [Prompt Registry](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/prompt-registry).

Reviewer: ```suggestion
You can define template content within the templating configuration, or refer to a template defined in the [Prompt Registry](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/prompt-registry).
```
